working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+
+&emsp;(B) &lArr; (A) &rArr; (C &hArr; D)
+
+Then, for each stage of inference, we find all the strongly connected components
+with no graph edges pointing to them, and remove them from the graph.  We repeat
+the process for successive stages, until no strongly connected components remain
+in the graph.  The set of arguments we will recursively infer during each stage
+is the union of the arguments in all the strongly connected components removed
+from the graph during that stage.
+
+In this example, that means that there will be two stages.  The first stage will
+consist of argument A, and the second stage will consist of arguments B, C, and
+D.
+
+(Note that the invariant mentioned earlier, that non-function literals are
+always placed in the first stage, is guaranteed by the fact that dependency
+analysis only draws an edge from _A_ to _B_ if _A_ is a function literal.)
+
+### What constitutes a function literal?
+
+The determination of whether an argument is a function literal is made
+syntactically.  An argument is a function literal if it is produced by one of
+the following grammar productions (as described in the language specification):
+
+* _\<functionExpression\>_
+
+* _\<namedArgument\>_, where the _\<expression\>_ part is a function literal.
+
+* `(` _\<expression\>_ `)`, where the _\<expression\>_ part is a function
+  literal.
+
+### Why is it safe to alter the visit order?
+
+Note that in most cases, flow analysis depends critically on the fact that type
+inference visits subexpressions in the same order they are evaluated.  For
+example, in the following code, the first argument to `f` requires a null check,
+because `int?` does not support `operator +`.
+
+```dart
+void f(int i, int j) { ... }
+void g(int? i) {
+  f(i! + 1, i = 0);
+}
+```
+
+If type inference instead visited the `i = 0` argument first, the `!` would be
+considered unnecessary (because `i = 0` promotes the type of `i` to non-nullable
+`int`), so we would be permitted to write `f(i + 1, i = 0)`.  However, this
+would fail at runtime, because arguments are evaluated in source order, so `i +
+1` is evaluated before `i = 0`.
+
+In general, we risk these sorts of breakages any time we change the order in
+which we perform type inference on invocation arguments.  However, there's an
+exception for arguments that are function literals: they may be safely
+re-ordered within the invocation, because the act of _evaluating_ a function
+literal doesn't actually have any runtime effect.  The runtime effect doesn't
+happen until the function literal is _called_, and the earliest this can
+possibly happen until after all the invocation arguments have been evaluated.
+
+This is why the invariant mentioned above (that non-function literals are always
+placed in the first stage) is critically important&mdash;it guarantees that the
+only arguments that will be type inferred out of order are function literals.
+
+### Why do we defer all function literals?
```

Comment:
This is the part that I think ought to be handled as a separate type inference improvement. Regardless of dependencies and argument ordering, it should be safe to postpone the introduction of the write-capturing effect of function literal arguments (that is, function literals that occur in an actual argument list and are not invoked during argument evaluation) until after the evaluation of all arguments.

If we do that separately then we keep things separate just because that is possible (which is an overall simplification), and then we don't need to consider a subdivision of the work in each stage (which turns out to be stage 0 only anyway, because that's the only one which can have both function literal arguments and other arguments).

I don't think it actually matters, but one part of the simplification would be that we could have, say, 3 stages, and then we could introduce some write-capture effects after stage 0, some other ones after stage 1, and finally some more after stage 2.

If we handle this aspect separately then we'd just gather the write-capture effects along the way and perform all of them after having handled the entire argument list.

Reply:
As I mentioned earlier, I prefer keeping the two improvements in the same proposal, both because the implementation is nearly complete (so there's no benefit in terms of implementation effort to splitting it up), and because the current behavior is poorly specified (so we can't really rewrite this into two separate smaller improvements without making reference to a specification that doesn't exist).

Reply:
OK!